File: packages/weather/src/tradewinds/weather/_awc.py
"""AWC METAR transform — maps AWC JSON response to observation schema dict.

This is THE shared transform. Both the SDK and ingest worker import it.
Output dicts validate against specs/observation.json (additionalProperties: false).
"""

from __future__ import annotations

import math
import re
from datetime import UTC, datetime
from typing import Any

from tradewinds._internal._bounds import (
    MAX_RAW_METAR_LEN,
    MAX_VISIBILITY_MILES,
    MAX_WX_CODES_LEN,
    SKY_BASE_MAX_FT,
    SLP_MAX_MB,
    SLP_MIN_MB,
    STATION_CODE_RE,
    TEMP_MAX_C,
    TEMP_MIN_C,
    WIND_DIR_BOUNDS,
    WIND_GUST_MAX,
    WIND_SPEED_MAX,
    bounded_float,
    bounded_float_min,
    bounded_int,
)
from tradewinds._internal._convert import celsius_to_fahrenheit, hpa_to_inhg


def icao_to_station_code(icao: str) -> str:
    """Strip leading K for 4-letter CONUS ICAO codes."""
    upper = icao.strip().upper()
    if upper.startswith("K") and len(upper) == 4:
        return upper[1:]
    return upper


def parse_awc_visibility(vis: Any) -> float | None:
    """Parse AWC visibility: '10+', '1/2', '2 1/4', '3/4', plain numbers.

    Returns miles or None. Caps at 99.99.
    """
    if vis is None:
        return None

    s = str(vis)
    if s == "" or s == "null":
        return None

    # "10+" -> 10
    if s.endswith("+"):
        try:
            n = float(s[:-1])
        except (ValueError, OverflowError):
            return None
        if not math.isfinite(n):
            return None
        return min(n, MAX_VISIBILITY_MILES)

    # Mixed number: "1 1/2", "2 1/4"
    if " " in s and "/" in s:
        parts = s.split(" ", 1)
        if len(parts) != 2:
            return None
        frac_parts = parts[1].split("/")
        if len(frac_parts) != 2:
            return None
        try:
            w = float(parts[0])
            n = float(frac_parts[0])
            d = float(frac_parts[1])
        except (ValueError, OverflowError):
            return None
        if not (math.isfinite(w) and math.isfinite(n) and math.isfinite(d) and d != 0):
            return None
        return min(w + n / d, MAX_VISIBILITY_MILES)

    # Simple fraction: "1/2", "1/4", "3/4", or "M1/4" (below-quarter-mile
    # AWC/METAR convention — codex review W3A P2). The leading 'M' means
    # "less than", which we represent as the same fractional value (the
    # observation schema treats this as the visibility value, not a flag).
    if "/" in s:
        if s.startswith("M") or s.startswith("m"):
            s = s[1:]
        frac_parts = s.split("/")
        if len(frac_parts) != 2:
            return None
        try:
            n = float(frac_parts[0])
            d = float(frac_parts[1])
        except (ValueError, OverflowError):
            return None
        if not (math.isfinite(n) and math.isfinite(d) and d != 0):
            return None
        return min(n / d, MAX_VISIBILITY_MILES)

    # Plain number
    try:
        n = float(s)
    except (ValueError, OverflowError):
        return None
    if not math.isfinite(n):
        return None
    return min(n, MAX_VISIBILITY_MILES)


def map_cloud_cover(cover: str | None) -> str | None:
    """Map AWC cloud cover code to standard abbreviation."""
    if cover is None:
        return None
    upper = cover.upper()
    if upper in ("CLR", "SKC", "FEW", "SCT", "BKN", "OVC", "VV"):
        return upper
    if upper == "CAVOK":
        return "CLR"
    return None


def _cloud_layer(layer: Any) -> tuple[str | None, int | None]:
    """Extract cover and base from a cloud layer dict. Safe against non-dict entries."""
    if not isinstance(layer, dict):
        return None, None
    base = bounded_int(_safe_int(layer.get("base")), 0, SKY_BASE_MAX_FT)
    return map_cloud_cover(layer.get("cover")), base


def _safe_int(v: Any) -> int | None:
    """Convert to int. Returns None on bad input."""
    if v is None:
        return None
    try:
        f = float(v)
        if not math.isfinite(f):
            return None
        return round(f)
    except (ValueError, TypeError, OverflowError):
        return None


def _safe_float(v: Any) -> float | None:
    """Convert to float. Returns None on bad input."""
    if v is None:
        return None
    try:
        f = float(v)
        return f if math.isfinite(f) else None
    except (ValueError, TypeError, OverflowError):
        return None


def _safe_precip(v: Any) -> float | None:
    """Parse precipitation. Trace 'T' → 0.0, numeric passthrough, else None."""
    if v is None:
        return None
    if isinstance(v, str) and v.strip().upper() == "T":
        return 0.0
    return _safe_float(v)


_PK_WND_RE = re.compile(r"PK WND (\d{3})(\d{2,3})/(\d{4})")

# T-group in METAR remarks: T{s}{SSS}{s}{DDD}
# s=0 positive, s=1 negative. SSS/DDD = tenths of °C.
# Example: T02560167 → 25.6°C / 16.7°C. T10390061 → -3.9°C / 6.1°C.
_TGROUP_RE = re.compile(r"\bT([01])(\d{3})([01])(\d{3})\b")


def _parse_peak_wind(
    raw_metar: str | None,
) -> tuple[int | None, int | None, str | None]:
    """Parse PK WND from METAR remarks. Returns (dir, speed_kt, time_hhmm)."""
    if not raw_metar:
        return None, None, None
    match = _PK_WND_RE.search(raw_metar)
    if not match:
        return None, None, None
    direction = int(match.group(1))
    speed = int(match.group(2))
    time_hhmm = match.group(3)
    if not (0 <= direction <= 360) or speed < 0:
        return None, None, None
    return direction, speed, time_hhmm


def _parse_tgroup(raw_metar: str | None) -> tuple[float | None, float | None]:
    """Parse T-group from METAR remarks for tenths-precision temperature.

    ASOS stations always include T-group in remarks. Format: T{s}{SSS}{s}{DDD}
    where s=0 positive, s=1 negative, SSS=temp tenths °C, DDD=dewpoint tenths °C.
    Searches only the remarks section (after RMK) to avoid false positives.
    Returns (temp_c, dewpoint_c) or (None, None) if not found.
    """
    if not raw_metar:
        return None, None
    # T-group is a remarks-only element — search only after RMK.
    # No RMK section = no T-group. Do NOT fallback to full string
    # to avoid false positives on body group patterns.
    rmk_idx = raw_metar.find("RMK")
    if rmk_idx < 0:
        return None, None
    match = _TGROUP_RE.search(raw_metar[rmk_idx:])
    if not match:
        return None, None
    t_sign = -1 if match.group(1) == "1" else 1
    t_val = int(match.group(2)) / 10.0 * t_sign
    d_sign = -1 if match.group(3) == "1" else 1
    d_val = int(match.group(4)) / 10.0 * d_sign
    return t_val, d_val


def awc_to_observation(m: dict[str, Any]) -> dict[str, Any] | None:
    """Convert a parsed AWC METAR dict to an observation schema dict.

    Returns None if icaoId or obsTime is invalid.
    Output matches specs/observation.json (no extra fields).
    """
    icao_id = m.get("icaoId")
    if not isinstance(icao_id, str) or not icao_id:
        return None

    obs_time = m.get("obsTime")
    if not isinstance(obs_time, (int, float)):
        return None

    station_code = icao_to_station_code(icao_id)
    if not STATION_CODE_RE.match(station_code):
        return None

    try:
        dt = datetime.fromtimestamp(obs_time, tz=UTC)
    except (OSError, OverflowError, ValueError):
        return None
    if not (1970 <= dt.year <= 2100):
        return None
    observed_at = dt.strftime("%Y-%m-%dT%H:%M:%SZ")

    metar_type = (m.get("metarType") or "METAR").upper()
    observation_type = "SPECI" if metar_type == "SPECI" else "METAR"

    # Wind direction: handle "VRB" -> None, bounded [0, 360]
    wdir: int | None = None
    raw_wdir = m.get("wdir")
    if raw_wdir is not None:
        if isinstance(raw_wdir, (int, float)):
            wdir = bounded_int(int(raw_wdir), *WIND_DIR_BOUNDS)
        elif raw_wdir != "VRB":
            try:
                parsed = float(raw_wdir)
                if math.isfinite(parsed):
                    wdir = bounded_int(int(parsed), *WIND_DIR_BOUNDS)
            except (ValueError, TypeError):
                pass

    wspd = bounded_int(_safe_int(m.get("wspd")), 0, WIND_SPEED_MAX)
    wgst = bounded_int(_safe_int(m.get("wgst")), 0, WIND_GUST_MAX)

    # Altimeter: AWC altim is in hPa, convert to inHg (no rounding)
    altim = hpa_to_inhg(_safe_float(m.get("altim")))

    # Sea-level pressure (already in mb/hPa)
    slp = _safe_float(m.get("slp"))
    if slp is not None and not (SLP_MIN_MB <= slp <= SLP_MAX_MB):
        slp = None

    # Cloud layers (safe against non-dict entries)
    clouds = m.get("clouds") or []
    cov1, base1 = _cloud_layer(clouds[0]) if len(clouds) > 0 else (None, None)
    cov2, base2 = _cloud_layer(clouds[1]) if len(clouds) > 1 else (None, None)
    cov3, base3 = _cloud_layer(clouds[2]) if len(clouds) > 2 else (None, None)
    cov4, base4 = _cloud_layer(clouds[3]) if len(clouds) > 3 else (None, None)

    # Raw METAR (truncate to 2048)
    raw_ob = m.get("rawOb")
    raw_metar: str | None = None
    if isinstance(raw_ob, str):
        raw_metar = raw_ob[:MAX_RAW_METAR_LEN]

    # Weather codes
    raw_wx = m.get("wxString")
    weather_codes: str | None = None
    if isinstance(raw_wx, str):
        weather_codes = raw_wx[:MAX_WX_CODES_LEN]

    # Temperature: T-group (tenths precision) overrides body group (whole degree).
    # ASOS always includes T-group in remarks. If present, use it.
    # Note: KNYC (Central Park) is NOT an ASOS station — may lack T-group,
    # falling back to whole-degree body group temps from AWC.
    temp_c = _safe_float(m.get("temp"))
    dewp_c = _safe_float(m.get("dewp"))
    tgroup_temp, tgroup_dewp = _parse_tgroup(raw_metar)
    if tgroup_temp is not None:
        temp_c = tgroup_temp
    if tgroup_dewp is not None:
        dewp_c = tgroup_dewp
    temp_c = bounded_float(temp_c, TEMP_MIN_C, TEMP_MAX_C, field="temp_c")
    dewp_c = bounded_float(dewp_c, TEMP_MIN_C, TEMP_MAX_C, field="dewpoint_c")
    temp_f = celsius_to_fahrenheit(temp_c)
    dewpoint_f = celsius_to_fahrenheit(dewp_c)

    # Peak wind from METAR remarks (PK WND dddss/hhmm), bounded
    pk_dir, pk_spd, pk_time = _parse_peak_wind(raw_metar)
    pk_dir = bounded_int(pk_dir, *WIND_DIR_BOUNDS)
    pk_spd = bounded_int(pk_spd, 0, WIND_GUST_MAX)

    # Precipitation (AWC provides 'precip' field; 'T' = trace → 0.0, non-negative)
    precip = bounded_float_min(_safe_precip(m.get("precip")), 0.0)

    # QC field bitmask
    qc_raw = m.get("qcField")
    qc_field = _safe_int(qc_raw)

    return {
        "station_code": station_code,
        "observed_at": observed_at,
        "observation_type": observation_type,
        "source": "awc",
        "temp_c": temp_c,
        "dewpoint_c": dewp_c,
        "temp_f": temp_f,
        "dewpoint_f": dewpoint_f,
        "wind_dir_degrees": wdir,
        "wind_speed_kt": wspd,
        "wind_gust_kt": wgst,
        "altimeter_inhg": altim,
        "sea_level_pressure_mb": slp,
        "sky_cover_1": cov1,
        "sky_base_1_ft": base1,
        "sky_cover_2": cov2,
        "sky_base_2_ft": base2,
        "sky_cover_3": cov3,
        "sky_base_3_ft": base3,
        "sky_cover_4": cov4,
        "sky_base_4_ft": base4,
        "visibility_miles": parse_awc_visibility(m.get("visib")),
        "weather_codes": weather_codes,
        "precip_1hr_inches": precip,
        "peak_wind_gust_kt": pk_spd,
        "peak_wind_dir": pk_dir,
        "peak_wind_time": pk_time,
        "snow_depth_inches": None,  # not available from AWC
        "qc_field": qc_field,
        "raw_metar": raw_metar,
    }

File: packages-ts/weather/src/_parsers/awc.ts
// AWC METAR parser — maps raw AWC JSON record to `Observation` row.
//
// Ported byte-faithfully from
// `packages/weather/src/tradewinds/weather/_awc.py::awc_to_observation`.
//
// Output `source` is `"awc.live"` — the canonical source-id used by
// downstream merge / cache layers (the Python parser emits `"awc"` and
// gets normalized to `"awc.live"` later; we emit the normalized form
// directly to match the TS-SDK source-id contract).
//
// =============================================================================
// Inlined bounds + conversion helpers
// =============================================================================
// `@tradewinds/core` exposes these via `internal/bounds.ts` and `internal/convert.ts`
// but the package only re-exports `fetchWithRetry` at its public surface
// (other helpers are documented as "deep imports for fetchers + parsers in
// later waves" but the `exports` field gates subpath access). Per the
// TS-W1 Wave 3 task scope ("All code in `packages-ts/weather/src/_fetchers/`
// + `_parsers/` ONLY. Do NOT modify any other package"), the needed
// constants and pure-function helpers are inlined verbatim here. Drift is
// caught by the parity-fixture gate in TS-W2.
// =============================================================================

import type { AwcMetarRaw } from "../_fetchers/awc.js";

// ---------------------------------------------------------------------------
// Bounds (mirror packages-ts/core/src/internal/bounds.ts)
// ---------------------------------------------------------------------------

const SLP_MIN_MB = 870.0;
const SLP_MAX_MB = 1084.0;
const TEMP_MIN_C = -90.0;
const TEMP_MAX_C = 60.0;
const MAX_RAW_METAR_LEN = 2048;
const MAX_WX_CODES_LEN = 256;
const MAX_VISIBILITY_MILES = 99.99;
const WIND_DIR_LO = 0;
const WIND_DIR_HI = 360;
const WIND_SPEED_MAX = 200;
const WIND_GUST_MAX = 250;
const SKY_BASE_MAX_FT = 60000;
const STATION_CODE_RE = /^[A-Z]{3,4}$/;

// ---------------------------------------------------------------------------
// Conversions (mirror packages-ts/core/src/internal/convert.ts)
// ---------------------------------------------------------------------------

const HPA_TO_INHG = 0.0295299875;

function finiteOrNull(v: number | null | undefined): number | null {
  if (v === null || v === undefined) return null;
  return Number.isFinite(v) ? v : null;
}

function celsiusToFahrenheit(c: number | null): number | null {
  const v = finiteOrNull(c);
  return v === null ? null : (v * 9) / 5 + 32;
}

function hpaToInhg(hpa: number | null): number | null {
  const v = finiteOrNull(hpa);
  return v === null ? null : v * HPA_TO_INHG;
}

// ---------------------------------------------------------------------------
// Bounded numerics (mirror bounds.ts helpers)
// ---------------------------------------------------------------------------

function boundedInt(val: number | null, lo: number, hi: number): number | null {
  if (val === null || val === undefined) return null;
  if (!Number.isFinite(val)) return null;
  return val >= lo && val <= hi ? val : null;
}

function boundedFloat(val: number | null, lo: number, hi: number): number | null {
  if (val === null || val === undefined) return null;
  if (!Number.isFinite(val)) return null;
  return val >= lo && val <= hi ? val : null;
}

function boundedFloatMin(val: number | null, lo: number): number | null {
  if (val === null || val === undefined) return null;
  if (!Number.isFinite(val)) return null;
  return val >= lo ? val : null;
}

// ---------------------------------------------------------------------------
// Observation row schema (matches specs/observation.json field set)
// ---------------------------------------------------------------------------

/**
 * Single observation row, matching `specs/observation.json`.
 *
 * Notes on null vs unknown:
 *  - Every field defaults to `null` if the upstream record omits it,
 *    fails bounds, or fails type parsing. This mirrors the Python
 *    parser (no exceptions on bad input — only on missing required keys).
 *  - `source` is always the string literal `"awc.live"` for AWC-sourced rows.
 *  - `observed_at` is ISO 8601 UTC with `Z` suffix.
 */
export interface Observation {
  readonly station_code: string;
  readonly observed_at: string;
  readonly observation_type: "METAR" | "SPECI";
  readonly source: "awc.live";
  readonly temp_c: number | null;
  readonly dewpoint_c: number | null;
  readonly temp_f: number | null;
  readonly dewpoint_f: number | null;
  readonly wind_dir_degrees: number | null;
  readonly wind_speed_kt: number | null;
  readonly wind_gust_kt: number | null;
  readonly altimeter_inhg: number | null;
  readonly sea_level_pressure_mb: number | null;
  readonly sky_cover_1: string | null;
  readonly sky_base_1_ft: number | null;
  readonly sky_cover_2: string | null;
  readonly sky_base_2_ft: number | null;
  readonly sky_cover_3: string | null;
  readonly sky_base_3_ft: number | null;
  readonly sky_cover_4: string | null;
  readonly sky_base_4_ft: number | null;
  readonly visibility_miles: number | null;
  readonly weather_codes: string | null;
  readonly precip_1hr_inches: number | null;
  readonly peak_wind_gust_kt: number | null;
  readonly peak_wind_dir: number | null;
  readonly peak_wind_time: string | null;
  readonly snow_depth_inches: number | null;
  readonly qc_field: number | null;
  readonly raw_metar: string | null;
}

// ---------------------------------------------------------------------------
// Public helpers
// ---------------------------------------------------------------------------

/**
 * Strip the leading `K` from 4-letter CONUS ICAO codes to get the
 * 3-letter NWS station code (`KNYC` → `NYC`). Non-K-prefixed codes
 * pass through unchanged after `.toUpperCase()`.
 */
export function icaoToStationCode(icao: string): string {
  const upper = icao.trim().toUpperCase();
  if (upper.startsWith("K") && upper.length === 4) {
    return upper.slice(1);
  }
  return upper;
}

/**
 * Map AWC cloud-cover token to the standard abbreviation set.
 * - Accepts {CLR, SKC, FEW, SCT, BKN, OVC, VV}
 * - `CAVOK` → `CLR`
 * - everything else (incl. null/undefined) → null
 */
export function mapCloudCover(cover: string | null | undefined): string | null {
  if (cover === null || cover === undefined) return null;
  const upper = String(cover).toUpperCase();
  if (
    upper === "CLR" ||
    upper === "SKC" ||
    upper === "FEW" ||
    upper === "SCT" ||
    upper === "BKN" ||
    upper === "OVC" ||
    upper === "VV"
  ) {
    return upper;
  }
  if (upper === "CAVOK") {
    return "CLR";
  }
  return null;
}

/**
 * Parse AWC visibility — handles all forms documented in the Python parser:
 *   - numeric (`10`)              → `10`
 *   - "10+"                       → `10`  (trailing plus = "or more")
 *   - "1/2"                       → `0.5`
 *   - "2 1/4"                     → `2.25`
 *   - "M1/4"                      → `0.25` (METAR "less than" prefix)
 *   - bad input / empty / "null"  → `null`
 *
 * All results are clamped at `MAX_VISIBILITY_MILES` (99.99).
 */
export function parseAwcVisibility(vis: string | number | null | undefined): number | null {
  if (vis === null || vis === undefined) return null;

  // Numeric pass-through (must be finite).
  if (typeof vis === "number") {
    if (!Number.isFinite(vis)) return null;
    return Math.min(vis, MAX_VISIBILITY_MILES);
  }

  const s = String(vis);
  if (s === "" || s === "null") return null;

  // "10+" → 10
  if (s.endsWith("+")) {
    const n = Number(s.slice(0, -1));
    if (!Number.isFinite(n)) return null;
    return Math.min(n, MAX_VISIBILITY_MILES);
  }

  // Mixed number "2 1/4" → 2.25  (space + slash both present)
  if (s.includes(" ") && s.includes("/")) {
    const parts = s.split(" ");
    if (parts.length !== 2) return null;
    const whole = Number(parts[0]);
    const frac = (parts[1] as string).split("/");
    if (frac.length !== 2) return null;
    const num = Number(frac[0]);
    const den = Number(frac[1]);
    if (!(Number.isFinite(whole) && Number.isFinite(num) && Number.isFinite(den) && den !== 0)) {
      return null;
    }
    return Math.min(whole + num / den, MAX_VISIBILITY_MILES);
  }

  // Simple fraction "1/2" or "M1/4"
  if (s.includes("/")) {
    let trimmed = s;
    if (trimmed.startsWith("M") || trimmed.startsWith("m")) {
      trimmed = trimmed.slice(1);
    }
    const frac = trimmed.split("/");
    if (frac.length !== 2) return null;
    const num = Number(frac[0]);
    const den = Number(frac[1]);
    if (!(Number.isFinite(num) && Number.isFinite(den) && den !== 0)) return null;
    return Math.min(num / den, MAX_VISIBILITY_MILES);
  }

  // Plain numeric string.
  const n = Number(s);
  if (!Number.isFinite(n)) return null;
  return Math.min(n, MAX_VISIBILITY_MILES);
}

// ---------------------------------------------------------------------------
// METAR remarks parsers (peak wind + T-group)
// ---------------------------------------------------------------------------

// `PK WND dddss/hhmm` (direction may be 2-3 digits; AWC payloads use 3+2)
const PK_WND_RE = /PK WND (\d{3})(\d{2,3})\/(\d{4})/;

// T-group in METAR remarks: T{s}{SSS}{s}{DDD}
// s=0 positive, s=1 negative. SSS/DDD = tenths of °C.
const TGROUP_RE = /\bT([01])(\d{3})([01])(\d{3})\b/;

function parsePeakWind(rawMetar: string | null): {
  dir: number | null;
  speed: number | null;
  time: string | null;
} {
  if (!rawMetar) return { dir: null, speed: null, time: null };
  const m = PK_WND_RE.exec(rawMetar);
  if (!m) return { dir: null, speed: null, time: null };
  const dir = Number.parseInt(m[1] as string, 10);
  const spd = Number.parseInt(m[2] as string, 10);
  const time = m[3] as string;
  if (!(dir >= 0 && dir <= 360) || spd < 0) {
    return { dir: null, speed: null, time: null };
  }
  return { dir, speed: spd, time };
}

function parseTGroup(rawMetar: string | null): { tempC: number | null; dewpC: number | null } {
  if (!rawMetar) return { tempC: null, dewpC: null };
  // T-group is a remarks-only element — search only after RMK to avoid
  // false positives on body group patterns.
  const rmkIdx = rawMetar.indexOf("RMK");
  if (rmkIdx < 0) return { tempC: null, dewpC: null };
  const m = TGROUP_RE.exec(rawMetar.slice(rmkIdx));
  if (!m) return { tempC: null, dewpC: null };
  const tSign = m[1] === "1" ? -1 : 1;
  const tVal = (Number.parseInt(m[2] as string, 10) / 10.0) * tSign;
  const dSign = m[3] === "1" ? -1 : 1;
  const dVal = (Number.parseInt(m[4] as string, 10) / 10.0) * dSign;
  return { tempC: tVal, dewpC: dVal };
}

// ---------------------------------------------------------------------------
// Safe numeric coercion (mirror Python `_safe_int`, `_safe_float`, `_safe_precip`)
// ---------------------------------------------------------------------------

function safeInt(v: unknown): number | null {
  if (v === null || v === undefined) return null;
  const f = typeof v === "number" ? v : Number(v);
  if (!Number.isFinite(f)) return null;
  // Python uses banker's `round()`; for non-negative half-values the
  // difference would matter, but observation payloads contain integral
  // counts (wind speed, sky base feet, qc bitmask). Use Math.round which
  // matches "round half away from zero" — adequate for the integer
  // domains we coerce here.
  return Math.round(f);
}

function safeFloat(v: unknown): number | null {
  if (v === null || v === undefined) return null;
  const f = typeof v === "number" ? v : Number(v);
  return Number.isFinite(f) ? f : null;
}

function safePrecip(v: unknown): number | null {
  if (v === null || v === undefined) return null;
  if (typeof v === "string" && v.trim().toUpperCase() === "T") {
    // Python returns 0.0 for trace; the task spec mentions "0.005 (Python
    // trace convention)" but the actual Python source maps "T" → 0.0 (see
    // `_safe_precip` in `_awc.py`). We match Python source byte-faithfully.
    return 0.0;
  }
  return safeFloat(v);
}

function cloudLayer(layer: unknown): { cover: string | null; base: number | null } {
  if (layer === null || typeof layer !== "object") return { cover: null, base: null };
  const obj = layer as { cover?: unknown; base?: unknown };
  const base = boundedInt(safeInt(obj.base), 0, SKY_BASE_MAX_FT);
  return { cover: mapCloudCover(obj.cover as string | null | undefined), base };
}

// ---------------------------------------------------------------------------
// Main parser
// ---------------------------------------------------------------------------

/**
 * Convert one raw AWC METAR record into the canonical observation row.
 *
 * Returns `null` if the record is missing `icaoId` or `obsTime`, or if the
 * station code can't be resolved via `icaoToStationCode` + STATION_CODE_RE,
 * or if `obsTime` produces an out-of-range date.
 *
 * Otherwise returns a fully-typed `Observation` with every field either a
 * validated value or `null`. Never throws.
 */
export function awcToObservation(m: AwcMetarRaw): Observation | null {
  // --- Required keys -----------------------------------------------------
  const icaoId = m.icaoId;
  if (typeof icaoId !== "string" || icaoId === "") return null;

  const obsTime = m.obsTime;
  if (typeof obsTime !== "number" || !Number.isFinite(obsTime)) return null;

  const stationCode = icaoToStationCode(icaoId);
  if (!STATION_CODE_RE.test(stationCode)) return null;

  // --- observed_at -------------------------------------------------------
  const dt = new Date(obsTime * 1000);
  if (Number.isNaN(dt.getTime())) return null;
  const year = dt.getUTCFullYear();
  if (!(year >= 1970 && year <= 2100)) return null;
  // ISO 8601 with second precision + "Z" suffix (matches Python
  // `%Y-%m-%dT%H:%M:%SZ`). `Date.toISOString` emits ms which we strip.
  const observedAt = `${dt.toISOString().slice(0, 19)}Z`;

  // --- observation_type --------------------------------------------------
  const metarType = typeof m.metarType === "string" ? m.metarType.toUpperCase() : "METAR";
  const observationType: "METAR" | "SPECI" = metarType === "SPECI" ? "SPECI" : "METAR";

  // --- Wind direction ----------------------------------------------------
  let wdir: number | null = null;
  const rawWdir = m.wdir;
  if (rawWdir !== null && rawWdir !== undefined) {
    if (typeof rawWdir === "number") {
      wdir = boundedInt(Math.trunc(rawWdir), WIND_DIR_LO, WIND_DIR_HI);
    } else if (rawWdir !== "VRB") {
      const parsed = Number(rawWdir);
      if (Number.isFinite(parsed)) {
        wdir = boundedInt(Math.trunc(parsed), WIND_DIR_LO, WIND_DIR_HI);
      }
    }
    // "VRB" → wdir stays null (Python leaves it None; task spec says
    // "0 (matches Python convention)" but the Python source on disk
    // leaves it None — we follow the source-of-truth).
  }

  // --- Wind speed / gust -------------------------------------------------
  const wspd = boundedInt(safeInt(m.wspd), 0, WIND_SPEED_MAX);
  const wgst = boundedInt(safeInt(m.wgst), 0, WIND_GUST_MAX);

  // --- Altimeter (hPa input → inHg output) -------------------------------
  const altim = hpaToInhg(safeFloat(m.altim));

  // --- Sea-level pressure (already mb/hPa) -------------------------------
  let slp = safeFloat(m.slp);
  if (slp !== null && !(slp >= SLP_MIN_MB && slp <= SLP_MAX_MB)) {
    slp = null;
  }

  // --- Cloud layers (up to 4) --------------------------------------------
  const clouds = m.clouds ?? [];
  const c0 = clouds[0] !== undefined ? cloudLayer(clouds[0]) : { cover: null, base: null };
  const c1 = clouds[1] !== undefined ? cloudLayer(clouds[1]) : { cover: null, base: null };
  const c2 = clouds[2] !== undefined ? cloudLayer(clouds[2]) : { cover: null, base: null };
  const c3 = clouds[3] !== undefined ? cloudLayer(clouds[3]) : { cover: null, base: null };

  // --- Raw METAR (truncate) ----------------------------------------------
  let rawMetar: string | null = null;
  if (typeof m.rawOb === "string") {
    rawMetar = m.rawOb.slice(0, MAX_RAW_METAR_LEN);
  }

  // --- Weather codes (truncate) ------------------------------------------
  let weatherCodes: string | null = null;
  if (typeof m.wxString === "string") {
    weatherCodes = m.wxString.slice(0, MAX_WX_CODES_LEN);
  }

  // --- Temperature + dewpoint (T-group overrides body group) -------------
  let tempC = safeFloat(m.temp);
  let dewpC = safeFloat(m.dewp);
  const { tempC: tgTemp, dewpC: tgDewp } = parseTGroup(rawMetar);
  if (tgTemp !== null) tempC = tgTemp;
  if (tgDewp !== null) dewpC = tgDewp;
  tempC = boundedFloat(tempC, TEMP_MIN_C, TEMP_MAX_C);
  dewpC = boundedFloat(dewpC, TEMP_MIN_C, TEMP_MAX_C);
  const tempF = celsiusToFahrenheit(tempC);
  const dewpointF = celsiusToFahrenheit(dewpC);

  // --- Peak wind ---------------------------------------------------------
  const pk = parsePeakWind(rawMetar);
  const pkDir = boundedInt(pk.dir, WIND_DIR_LO, WIND_DIR_HI);
  const pkSpd = boundedInt(pk.speed, 0, WIND_GUST_MAX);

  // --- Precip + QC -------------------------------------------------------
  const precip = boundedFloatMin(safePrecip(m.precip), 0.0);
  const qcField = safeInt(m.qcField);

  return {
    station_code: stationCode,
    observed_at: observedAt,
    observation_type: observationType,
    source: "awc.live",
    temp_c: tempC,
    dewpoint_c: dewpC,
    temp_f: tempF,
    dewpoint_f: dewpointF,
    wind_dir_degrees: wdir,
    wind_speed_kt: wspd,
    wind_gust_kt: wgst,
    altimeter_inhg: altim,
    sea_level_pressure_mb: slp,
    sky_cover_1: c0.cover,
    sky_base_1_ft: c0.base,
    sky_cover_2: c1.cover,
    sky_base_2_ft: c1.base,
    sky_cover_3: c2.cover,
    sky_base_3_ft: c2.base,
    sky_cover_4: c3.cover,
    sky_base_4_ft: c3.base,
    visibility_miles: parseAwcVisibility(m.visib),
    weather_codes: weatherCodes,
    precip_1hr_inches: precip,
    peak_wind_gust_kt: pkSpd,
    peak_wind_dir: pkDir,
    peak_wind_time: pk.time,
    snow_depth_inches: null, // AWC doesn't carry snow-depth in METARs
    qc_field: qcField,
    raw_metar: rawMetar,
  };
}
